gusto/time_discretisation.py

```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
+        solver.solve()
+        x_out.assign(self.x_out)
+
+
+class AdamsBashforth(MultilevelTimeDiscretisation):
+    """
+    Implements the explicit multistep Adams-Bashforth timestepping method of general order up to 5
+
+    The general AB timestepping method for operator F is written as
+    y^(n+1) = y^n + dt*(b_0*F[y^(n)] + b_1*F[y^(n-1)] + b_2*F[y^(n-2)] + b_3*F[y^(n-3)] + b_4*F[y^(n-4)])
+    """
+    def __init__(self, domain, order, field_name=None,
+                 solver_parameters=None, options=None):
+        """
+        Args:
+            domain (:class:`Domain`): the model's domain object, containing the
+                mesh and the compatible function spaces.
+            field_name (str, optional): name of the field to be evolved.
+                Defaults to None.
+            order (float, optional): order of scheme
+            solver_parameters (dict, optional): dictionary of parameters to
+                pass to the underlying solver. Defaults to None.
+            options (:class:`AdvectionOptions`, optional): an object containing
+                options to either be passed to the spatial discretisation, or
+                to control the "wrapper" methods, such as Embedded DG or a
+                recovery method. Defaults to None.
+
+        Raises:
+            ValueError: if order is not provided, or is in incorrect range.
+        """
+
+        if (order > 5 or order < 1):
+            raise ValueError("Adams-Bashforth of order greater than 5 not implemented")
+        if isinstance(options, (EmbeddedDGOptions, RecoveryOptions)):
+            raise NotImplementedError("Only SUPG advection options have been implemented for this time discretisation")
+        if not solver_parameters:
+            # theta method leads to asymmetric matrix, per lhs function below,
+            # so don't use CG
+            solver_parameters = {'ksp_type': 'gmres',
+                                 'pc_type': 'bjacobi',
+                                 'sub_pc_type': 'ilu'}
+
+        super().__init__(domain, field_name,
+                         solver_parameters=solver_parameters,
+                         options=options)
+
+        self.order = order
+
+    def setup(self, equation, uadv=None, apply_bcs=True, *active_labels):
+        super().setup(equation=equation, uadv=uadv, apply_bcs=apply_bcs,
+                      *active_labels)
+
+        self.x = [Function(self.fs) for i in range(self.nlevels)]
+
+        if (self.order == 1):
+            self.b = [1.0]
+        elif (self.order == 2):
+            self.b = [-(1.0/2.0), (3.0/2.0)]
+        elif (self.order == 3):
+            self.b = [(5.0)/(12.0), -(16.0)/(12.0), (23.0)/(12.0)]
+        elif (self.order == 4):
+            self.b = [-(9.0)/(24.0), (37.0)/(24.0), -(59.0)/(24.0), (55.0)/(24.0)]
+        elif (self.order == 5):
+            self.b = [(251.0)/(720.0), -(1274.0)/(720.0), (2616.0)/(720.0), -(2774.0)/(720.0), (2901.0)/(720.0)]
```

Comment:
I will presume that all of these coefficients are correct! Did you manage to test the convergence somewhere?

Reply:
For this and previous comment: I believe so, I will do some convergence tests for all schemes to check and post here